File: leetcode/Stack/20_valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        p = {')': '(', '}': '{', ']': '['}
        for c in s:
            if c in p.values():
                stack.append(c)
            elif not stack or stack.pop() != p[c]:
                return False

        return stack == []

File: ts/leetcode/20. Valid Parentheses.ts
function isValid(s: string): boolean {
    const stack: string[] = [];
    const mapping: { [index: string]: string } = {
        '}': '{',
        ']': '[',
        ')': '(',
    };
    for (let i = 0; i < s.length; i++) {
        const c = s[i];
        if (mapping[c]) {
            // right parentheses
            if (stack && stack[stack.length - 1] === mapping[c]) {
                stack.pop();
            } else {
                return false;
            }
        } else {
            // left parentheses
            stack.push(c);
        }
    }
    return !stack.length;
}
